Use a custom std::char_traits<byte> for implementations lacking a generic implementation

tambry opened the pull request:
Not sure what style you'd prefer the docstrings in (`byte_string` or `pqxx::byte_string`).  

We could also name these `pqxx::bytes` and `pqxx::bytes_view` to allow switching to `std::vector<std::byte>` and `std::span<std::byte>` in the next major version that bumps the C++ requirement without ending up with incorrect names for them.

---
Standard does not require a generic implementation. libc++ has deprecated its in version 18 and will remove it in version 19.

Replace with type aliases that will use a custom char_traits for libc++ 18 and newer. This is made conditional for libc++ 18 and newer only as this is a source breaking change.

Fixes: #726

jtv commented:
Thank you for doing this!  And perfect timing, too.

It may take me a few days to look this over again.  I'll try not to bother you with petty details like how to spell the names in documentation — I can edit those myself later.

The names you suggest sound like a good idea, though I'll have to look around older libpqxx versions because I think at one point I had a different class for this called `pqxx::bytes`.

jtv commented:
Oh, some build errors:

`../include/pqxx/util.hxx:283:5: error: '_LIBCPP_VERSION' is not defined, evaluates to 0 [-Werror,-Wundef]`

That's because I like to build libpqxx with the strictest compiler options I can, so that build problems will be more likely to show up in my build than other people's.

I suppose these lines:
```cxx
#if _LIBCPP_VERSION >= 180000
```

...should look more like:
```cxx
#if defined(_LIBCPP_VERSION) && _LIBCPP_VERSION >= 180000
```

But perhaps checking the library version can be a little brittle as well?  Is there a better condition?

tambry commented:
> It may take me a few days to look this over again. I'll try not to bother you with petty details like how to spell the names in documentation — I can edit those myself later.

I'll be happy to fix things up as you desire to hopefully not require any noisy follow-up commits on your part.

> The names you suggest sound like a good idea, though I'll have to look around older libpqxx versions because I think at one point I had a different class for this called pqxx::bytes.

I've changed over the names to `pqxx::bytes` and `pqxx::bytes_view`.  
I like that they're much shorter and enable the future migration to `std::vector<std::byte>` and `std::span<std::byte>` without ending up with illogical names, doing another rename or having multiple aliases for the same type.

> But perhaps checking the library version can be a little brittle as well? Is there a better condition?

I didn't immediately see a way to do that, but thinking a little harder I've now managed to use SFINAE to make this generic and not specific to libc++. 🎉 

---

I've also significantly improved the commit message. 😉

jtv commented:
I'm sorry for being so slow — lots of drama.

Another small note: I actually have no idea what the rules are for multi-line strings in `NEWS`.  So keep them to one line per entry.  But if you know better, that's fine with me.  :-D

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+
+    std::byte *beginning{dst};
+
+    if (dst < src)
```
Does `std::memmove()` make use of these traits?  If not, then why not just call that?  It may include some clever optimisation.

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+  {
+    std::byte *beginning{dst};
+
+    for (; size; --size, ++dst, ++src) assign(*dst, *src);
```
And here...  Not just `std::memcpy()`?

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+
+template<typename TYPE>
+struct has_generic_char_traits<
+  TYPE, std::void_t<decltype(std::char_traits<TYPE>::eof)>> : std::true_type
```
It's not clear to me what this does...  Is this the SFINAE test for whether there is a traits type for `TYPE`?

Looks worth a comment!

tambry replied:
Precisely, SFINAE check for whether there's a `std::char_traits<TYPE>::eof()` available, i.e. `std::char_traits<TYPE>` is available. Will add a comment.

jtv replied:
Thanks!

tambry replied:
Well it's checking if an applicable specialisation exists with `eof()` inside it. The standard library will call it anyway so if it's just a declaration it'd fail to compile. I can't quite see how this break with any actual standard library. 🤔

tambry commented:
@jtv Basically the whole implementation of `byte_char_traits` is a functional-equivalent of libc++'s generic `std::char_traits`. I didn't think much about it when writing the original patch as I wanted to get it just compiling without any change in behaviour when I did the original patch locally. Will go over with actual thought now. 😅

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
 #  define PQXX_POTENTIAL_BINARY_ARG typename
 #endif
 
+struct byte_char_traits : std::char_traits<char>
```
I think when a reader is likely to wonder _why_ code does something, that's usually a good place for a comment.  Perhaps you can document this type?  I understand right now why it's there, but if I put myself in the place of someone trying to do maintenance on the library in the future, I think I'd be much happier if this type explained itself.

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+
+inline constexpr bool has_generic_bytes_char_traits =
+  has_generic_char_traits<std::byte>::value;
+using bytes = std::conditional<
```
These types _definitely_ need documentation, because we'll expect users to understand them.

In particular...  Do you think it would make sense to put a big warning in there saying "we may suddenly decide to replace this type with `std::vector<std::byte>` in the future, but we're not doing that right now for historical reasons"?

jtv commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+  std::basic_string<std::byte, byte_char_traits>>::type;
+using bytes_view = std::conditional<
+  has_generic_bytes_char_traits, std::basic_string_view<std::byte>,
+  std::basic_string_view<std::byte, byte_char_traits>>::type;
```
Looking a bit further ahead, would it make sense to reserve the right to replace this with `std::span<std::byte>` in the future?

jtv commented:
I've read through it now.  Once again thanks for doing this!  My review style dictates that I [always have notes](https://nigelb.me/2011-10-30-please-nitpick) on a patch...  Please don't be discouraged.  :-)

jtv commented:
Oh, you asked about docstring styles.  Same as everywhere else...
* Doxygen-style comments
* with Markdown-style markup
* on basically everything.

jtv commented:
Thanks for keeping me updated.  Looking forward to this.

I've reserved significant time for libpqxx over December.  Hopefully we can get a solid 7.9 ready for release, and lay more groundwork for 8.0 which will require C++20 as a minimum.  These are also two opportunities to make incompatible changes.

jtv commented:
@tambry I've got some other changes lined up, but when I think any would conflict with your work, I'll try to hold them until yours is merged.

jtv commented:
@tambry just a gentle reminder so you don't forget.  Or if you're at PGConf.eu, maybe we can just discuss it in person.  :-)

jtv commented:
I've got a few PRs waiting to be merged that should only conflict with this one on the NEWS file.  Perhaps I should go ahead and merge those - it'll be annoying for you, for which I apologise, but not complicated

tambry commented:
I'm very much familiar with rebasing and solving conflicts so won't be an issue. I should hopefully be able to finish this up over the holidays. Been really busy, sorry! 😅

jtv commented:
Excellent, I'll merge a few then.  Thanks for sticking with it, and...  merry Christmas!

jtv commented:
@tambry perhaps you can merge or rebase the latest upstream changes?

jtv commented:
@tambry I took the liberty of resolving conflicts in the Github web UI.  (Forgot you could do that.)

hope that doesn't conflict with any changes you may have waiting on your own system.  I feel a bit stuck — I'd love to release a 7.9 soon with this change on board, as well as extensive fixes to improve compliance with the C++ Core Guidelines etc.

The latter I _could_ merge first, but it may increase the conflict with any local changes that you may have pending.

jtv commented:
My apologies @tambry — it looks like I left a double declaration of `quote_raw()` in there while I was trying to fix the conflicts.

tambry commented:
@jtv Sorry for the lack of replies. My situation with not being caught up with both work and personal things remains the same, but I feel like I some hope of not getting more left behind. 😞  
I'll try to get around to this tomorrow as this has been buzzing in my head for at least 1.5 months by now.

tambry commented on include/pqxx/util.hxx:
```diff
@@ -280,18 +280,135 @@ struct PQXX_LIBEXPORT thread_safety_model
+  {
+    std::size_t length{};
+
+    for (; !eq(*data, std::byte{0}); ++data) ++length;
```
It **is** insanity! If you have a `std::basic_string` there _must_ be an usable `length()` as if it were a C-string. If you're just holding bytes it obviously makes no sense. But we still need _something_. This is equivalent to what standard libraries' generic implementations did before, so... 🤷‍♂️

tambry commented:
* Rebased
* Added comments
* Added avoidance for libc++ 18's deprecation warning
* Changed `byte_char_traits` to use `std::mem*()` and `std::strchr()` instead of implementing ourselves.

I've closed threads that I feel I dealt with sufficiently. Please do re-open in case of any questions or if you feel I resolved them inadequately.

jtv commented:
I guess it's my turn to apologise @tambry...  You did lots of work two weeks ago and I hadn't noticed!  Didn't see any notifications from githubfor some reason.  And of course I was traveling for FOSDEM etc.

Thanks again for doing this work.  I'm happy with the explanations and the changes, so I'd like to merge it.  I'll make some minor tweaks such as applying the Doxygen comment format, and I'll try to make sure the documentation is up to date on the change.

And of course I'd also like to make sure that in C++20, people will feel encouraged to use `std::span<std::byte>` and `std::vector<std::byte>` as well.  Concepts will just hide the differences.

jtv commented:
@tambry I think I have a solution for the `length()` problem: the function has to be declared but I can just not _define_ it!

jtv commented:
Absolutely!  I tried wearing my PGDay badge but it broke.
